include/lbann/utils/threads/type_erased_function.hpp

```diff
@@ -0,0 +1,91 @@
+#ifndef __LBANN_TYPE_ERASED_FUNCTION_HPP__
+#define __LBANN_TYPE_ERASED_FUNCTION_HPP__
+
+#include <type_traits>
+
+#include "memory.hpp"
+
+namespace lbann {
+
+/** \class type_erased_function
```

Comment:
The documentation of this class is remarkably complete. We should always do that. :P

Reply:
:-D